build.gradle

```diff
@@ -59,57 +59,93 @@ task configureGroovyAll {
     inputs.files configurations.groovy
 
     doFirst {
-        def extensionClasses = []
-        def staticExtensionClasses = []
+        def toMerge = [
+            'LICENSE'                       : 'META-INF/LICENSE',
+            'NOTICE'                        : 'META-INF/NOTICE',
+            'AST_TRANSFORMATION'            : 'META-INF/services/org.codehaus.groovy.transform.ASTTransformation',
+            'EXTENSION_MODULE'              : 'META-INF/groovy/org.codehaus.groovy.runtime.ExtensionModule',
+            'GROOVY_RELEASE_INFO_PROPERTIES': 'META-INF/groovy-release-info.properties',
+            'GROOVY_RUNNER'                 : 'META-INF/services/org.apache.groovy.plugin.GroovyRunner',
+            'INDEX_LIST'                    : 'META-INF/INDEX.LIST',
+            'MANIFEST_MF'                   : 'META-INF/MANIFEST.MF'
+        ]
+
+        def findDuplicates = { List<FileTree> fileForest ->
+            Map<String, List<File>> filesByRelativePath = [:].withDefault { [] }
+            fileForest*.visit([
+                visitDir : {},
+                visitFile: { filesByRelativePath[it.relativePath.toString()] << it.file }
+            ] as FileVisitor)
+            filesByRelativePath.findAll { it.value.size() > 1 }
+        }
 
-        tasks.groovyAll {
-            from writeVersionProperties
+        def concatenate = { List<File> files -> files*.text.join('\n') }
 
-            configurations.groovy.filter { it.name ==~ /groovy-.*\.jar/ }.each { groovyJar ->
-                def extractedFiles = zipTree(groovyJar)
+        def joinValues = { String properties, String keyName ->
+            (properties =~ /$keyName\s*=\s*(.+)/)*.getAt(1).join(',')
+        }
 
-                extractedFiles.filter { it.name == "org.codehaus.groovy.runtime.ExtensionModule" }.each { extensionModuleFile ->
-                    def extensionModule = new Properties()
-                    extensionModuleFile.withInputStream { input ->
-                        extensionModule.load(input)
-                    }
-                    extensionClasses << extensionModule.extensionClasses
-                    staticExtensionClasses << extensionModule.staticExtensionClasses
-                }
+        def mergeFiles = { String relative, List<File> files ->
+            switch (relative) {
+                case toMerge.LICENSE:
+                case toMerge.NOTICE:
+                case toMerge.AST_TRANSFORMATION:
+                    return concatenate(files)
+                case toMerge.GROOVY_RELEASE_INFO_PROPERTIES:
+                    return files[0].text
+                case toMerge.EXTENSION_MODULE:
```

Comment:
similar solution in original, merged Groovy 2.4: https://github.com/apache/groovy/blob/REL-BRANCH-2.4.0-GA/gradle/assemble.gradle#L269

Reply:
:+1: